### PR TITLE
[FEATURE] specVersion 2.1: Add support for "customConfiguration"

### PR DESCRIPTION
**Pull request by matz3**
- [x] Add "customConfiguration" object for arbitrary configuration
- [x] Also cover kind extension
- [x] Add tests for all types / kinds (maybe generic?)
- [x] Add generic tests to validate common configuration of 2.1 and 2.0 to avoid a lot of duplicated tests

### CODE REVIEW

File: test/lib/validation/schema/specVersion/2.0/kind/project/commonUtils.js [hunk 1]
```diff
@@ -0,0 +1,494 @@
+const test = require("ava");
+
+test("commonUtils", async (t) => {
```
Comment (matz3) [inline test/lib/validation/schema/specVersion/2.0/kind/project/commonUtils.js:3]: I think we should rather exclude this file in the ava config (see package.json).

File: test/lib/validation/schema/specVersion/2.0/kind/project/application.js [hunk 2]
```diff
@@ -34,956 +35,661 @@ test.after.always((t) => {
...
-	await assertValidation(t, {
-		"specVersion": "2.0",
-		"type": "application",
-		"metadata": {
-			"name": "my-application",
-			"deprecated": false
-		}
-	});
-});
+	test("framework configuration: Invalid (specVersion " + version + ")", async (t) => {
```
Comment (matz3) [inline test/lib/validation/schema/specVersion/2.0/kind/project/application.js:495]: I think all framework tests should be common across application, library and theme-library.

**Comment by matz3** [comment]
All types of kind `extension` should also have the `customConfiguration`

**Comment by RandomByte** [comment]
Resolves https://github.com/SAP/ui5-project/issues/304

File: test/lib/validation/schema/specVersion/2.0/kind/project/commonUtils.js (hunk 1 above)
Comment (tobiasso85) [reply test/lib/validation/schema/specVersion/2.0/kind/project/commonUtils.js:3]: good point! done

File: test/lib/validation/schema/specVersion/2.0/kind/project/application.js (hunk 2 above)
Comment (tobiasso85) [reply test/lib/validation/schema/specVersion/2.0/kind/project/application.js:495]: these are run for all kinds but not for all versions
(added comments)

File: test/lib/validation/schema/__helper__/customConfiguration.js [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+					message: "should NOT have additional properties",
+					params: {
+						additionalProperty: "customConfiguration",
+					},
+					schemaPath: "#/else/additionalProperties",
+				}
+			]);
+		});
+
+		test(type + " (specVersion 2.0): Additional property", async (t) => {
```
[the remark this answers is not in the document]
Comment (tobiasso85) [reply test/lib/validation/schema/__helper__/customConfiguration.js:39]: relocated the tests back to project/extension

File: test/lib/validation/schema/__helper__/customConfiguration.js [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+				keyword: "additionalProperties",
+				message: "should NOT have additional properties",
+				params: {
+					additionalProperty: "notAllowed",
+				},
+				schemaPath: "#/else/additionalProperties",
+			}]);
+		});
+
+		test(type + " (specVersion 2.1): Additional property", async (t) => {
```
[the remark this answers is not in the document]
Comment (tobiasso85) [reply test/lib/validation/schema/__helper__/customConfiguration.js:58]: relocated the tests back to project/extension

File: test/lib/validation/schema/__helper__/customConfiguration.js [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+				"metadata": {
+					"name": "my-" + type
+				},
+				"customConfiguration": {
+					"foo": "bar"
+				}
+			}, oMandatoryProperty));
+		});
+
+		test(type + ": Additional property (specVersion 2.1)", async (t) => {
```
[the remark this answers is not in the document]
Comment (tobiasso85) [reply test/lib/validation/schema/__helper__/customConfiguration.js:90]: removed

File: test/lib/validation/schema/__helper__/customConfiguration.js [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+ */
+module.exports = {
+	/**
+	 * Executes the tests for different kind of projects, e.g. "application", "library", "theme-library" and "module"
+	 *
+	 * @param {Function} test ava test
+	 * @param {Function} assertValidation assertion function
+	 * @param {string} type one of "project-shim", "server-middleware" "task",
+	 *  "application", "library", "theme-library" and "module"
+	 * @param {Object} oMandatoryProperty mandatory configuration content
```
[the remark this answers is not in the document]
Comment (tobiasso85) [reply test/lib/validation/schema/__helper__/customConfiguration.js:12]: renamed to additionalConfiguration

File: test/lib/validation/schema/__helper__/extensionUtils.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+	 * @param {string} type one of "project-shim", "server-middleware" and "task"
+	 * @param {Object} oMandatoryProperty mandatory configuration content
+	 */
+	runTests: function(test, assertValidation, type, oMandatoryProperty) {
+		oMandatoryProperty = oMandatoryProperty || {};
+		oMandatoryProperty = Object.assign({"kind": "extension"}, oMandatoryProperty);
+
+		customConfiguration.runTests(test, assertValidation, type, oMandatoryProperty);
+
+		["2.0", "2.1"].forEach((version) => {
```
[the remark this answers is not in the document]
Comment (tobiasso85) [reply test/lib/validation/schema/__helper__/extensionUtils.js:21]: done

File: test/lib/validation/schema/__helper__/extensionUtils.js [hunk 8]
```diff
@@ -0,0 +1,56 @@
+const customConfiguration = require("./customConfiguration");
```
[the remark this answers is not in the document]
Comment (tobiasso85) [reply test/lib/validation/schema/__helper__/extensionUtils.js:1]: removed "Utils"

**Comment by matz3** [comment]
Looks ok from my side. Maybe @RandomByte also want to have a look

File: test/lib/validation/schema/__helper__/extension.js [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+		additionalConfiguration = additionalConfiguration || {};
+		additionalConfiguration = Object.assign({"kind": "extension"}, additionalConfiguration);
+
+		customConfiguration.defineTests(test, assertValidation, type, additionalConfiguration);
+
+		["2.0", "2.1"].forEach((specVersion) => {
+			test(`kind: extension / type: ${type} basic (${specVersion})`, async (t) => {
+				await assertValidation(t, Object.assign({
+					"specVersion": specVersion,
+					"kind": "extension",
```
Comment (RandomByte) [inline test/lib/validation/schema/__helper__/extension.js:25]: Duplicate information. Already defined in `additionalConfiguration`, so should be removed here
```suggestion
```

File: test/lib/validation/schema/__helper__/extension.js [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+					"metadata": {
+						"name": "my-" + type
+					}
+				}, additionalConfiguration));
+			});
+
+			test(`kind: extension / type: ${type} additionalProperties (${specVersion})`, async (t) => {
+				await assertValidation(t, Object.assign({
+					"specVersion": specVersion,
+					"kind": "extension",
```
Comment (RandomByte) [inline test/lib/validation/schema/__helper__/extension.js:36]: Duplicate information. Already defined in `additionalConfiguration`, so should be removed here

```suggestion
```

File: test/lib/validation/schema/__helper__/extension.js [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+					keyword: "additionalProperties",
+					message: "should NOT have additional properties",
+					params: {
+						"additionalProperty": "resources"
+					},
+					schemaPath: specVersion === "2.1" ? "#/then/additionalProperties" : "#/else/additionalProperties"
+				}]);
+			});
+
+			test(`${type} Invalid configuration: Additional property (${specVersion})`, async (t) => {
```
Comment (RandomByte) [inline test/lib/validation/schema/__helper__/extension.js:53]:
```suggestion
			test(`kind: extension / type: ${type} Invalid configuration: Additional property (${specVersion})`, async (t) => {
```

Align with other test names

File: lib/validation/schema/specVersion/2.0/kind/extension/task.json [hunk 12]
```diff
@@ -3,25 +3,60 @@
...
+			"kind": {
+				"enum": ["extension"]
+			},
+			"type": {
+				"enum": ["task"]
+			},
+			"metadata": {
+				"$ref": "../extension.json#/definitions/metadata"
+			},
+			"task": {
```
[the remark this answers is not in the document]
Comment (matz3) [reply lib/validation/schema/specVersion/2.0/kind/extension/task.json:25]: Done

File: lib/validation/schema/specVersion/2.0/kind/extension/server-middleware.json [hunk 13]
```diff
@@ -3,25 +3,61 @@
...
+			"kind": {
+				"enum": ["extension"]
+			},
+			"type": {
+				"enum": ["server-middleware"]
+			},
+			"metadata": {
+				"$ref": "../extension.json#/definitions/metadata"
+			},
+			"middleware": {
```
[the remark this answers is not in the document]
Comment (matz3) [reply lib/validation/schema/specVersion/2.0/kind/extension/server-middleware.json:26]: Done

File: test/lib/validation/schema/__helper__/extension.js (hunk 9 above)
Comment (matz3) [reply test/lib/validation/schema/__helper__/extension.js:25]: Done

File: test/lib/validation/schema/__helper__/extension.js (hunk 10 above)
Comment (matz3) [reply test/lib/validation/schema/__helper__/extension.js:36]: Done

File: test/lib/validation/schema/__helper__/extension.js (hunk 11 above)
Comment (matz3) [reply test/lib/validation/schema/__helper__/extension.js:53]: Done

**Comment by RandomByte** [comment]
We forgot to enhance this if with the new spec version: https://github.com/SAP/ui5-project/blob/71499c022749b672a3cfdbaa4d583eb4f978676b/lib/projectPreprocessor.js#L279

Also tests were not adapted. Currently you can't use specVersion 2.1 in a project, it will raise an error

```
Unsupported specification version 2.1 defined for project openui5-sample-app. Your UI5 CLI installation might be outdated. For details see https://sap.github.io/ui5-tooling/pages/Configuration/#specification-versions
```

**Pull request merged**